File: rust/src/old/recursive_backtracking.rs
use crate::direction::Direction;
use crate::maze::Maze;
use rand::prelude::*;

/// Here’s the mile-high view of **recursive backtracking**:
///
/// 1. Choose a starting point in the field.
/// 2. Randomly choose a wall at that point and carve a passage through to the adjacent cell, but
///    only if the adjacent cell has not been visited yet. This becomes the new current cell.
/// 3. If all adjacent cells have been visited, back up to the last cell that has uncarved walls and repeat.
/// 4. The algorithm ends when the process has backed all the way up to the starting point.
///
/// cx,cy is the current cell being visited
fn carve_passage(cx: usize, cy: usize, maze: &mut Maze) {
    // choose a random direction
    let mut directions = [Direction::N, Direction::S, Direction::E, Direction::W];
    directions.shuffle(&mut thread_rng());

    for direction in directions.iter() {
        let nx = (cx as i8 + direction.dx()) as usize;
        let ny = (cy as i8 + direction.dy()) as usize;

        // new cell must be within the bounds of the grid and must not already be visited
        if maze.in_bounds(nx, ny) && maze[[ny, nx]] == 0 {
            // "carve" a passage from the current cell to the next cell
            maze[[cy, cx]] |= *direction as u8;
            maze[[ny, nx]] |= direction.opposite() as u8;

            carve_passage(nx, ny, maze);
        }
    }
}

pub fn recursive_backtracking(width: usize, height: usize) -> Maze {
    let mut maze = Maze::new(width, height);
    carve_passage(0, 0, &mut maze);
    maze
}

File: python/old/recursive_backtracking.py
from maze import (Maze, Direction)
import random
import sys

# Recursive Backtracking
# Here’s the mile-high view of **recursive backtracking**:
#
# 1. Choose a starting point in the field.
# 2. Randomly choose a wall at that point and carve a passage through to the adjacent cell, but
#    only if the adjacent cell has not been visited yet. This becomes the new current cell.
# 3. If all adjacent cells have been visited, back up to the last cell that has uncarved walls and repeat.
# 4. The algorithm ends when the process has backed all the way up to the starting point.
#    cx,cy is the current cell being visited


def recurse(cx, cy, maze: Maze):
    """
    generates a random maze on the passed in maze, using the recursive backtracing algorithm. The passed in maze in
    modified in place
    :param cx: cell x position to begin the algorithm at, 0 is the leftmost column of maze
    :param cy: cell y position to begin the algorithm at, 0 is the topmost row of the maze
    :param maze: the empty maze to use for the algorithm
    """
    directions = [Direction.N, Direction.S, Direction.E, Direction.W]
    random.shuffle(directions)

    for dir in directions:
        nx = cx + Direction.dx(dir)
        ny = cy + Direction.dy(dir)

        if 0 <= ny < len(maze.field) and nx >= 0 and nx < len(maze.field[ny]) and maze.field[ny][nx] == 0:
            maze.field[cy][cx] |= dir.value
            maze.field[ny][nx] |= Direction.opposite(dir).value
            recurse(nx, ny, maze)


def recursive_backtracking(start_x: int, start_y: int, maze_width: int, maze_height: int) -> Maze:
    """
    generates a random maze using recursive backtracking
    :param start_x: the x  position in the maze to start at
    :param start_y: the y position in the maze to start at
    :param maze_width: the width of the maze to generate
    :param maze_height: the height of the maze to generate
    :return: a random Maze
    """
    maze = Maze(maze_width,  maze_height)
    recurse(start_x, start_y, maze)
    return maze


width = int(sys.argv[1]) if len(sys.argv) > 1 else 20
height = int(sys.argv[2]) if len(sys.argv) > 2 else 20
maze = recursive_backtracking(0, 0, width, height)
maze.display()